adapters/marsmedia/marsmedia.go

```diff
@@ -0,0 +1,167 @@
+package marsmedia
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type MarsmediaAdapter struct {
+	URI string
+}
+
+func (a *MarsmediaAdapter) MakeRequests(requestIn *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	request := *requestIn
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}}
+	}
+
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(request.Imp[0].Ext, &bidderExt); err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}}
+	}
+
+	var marsmediaExt openrtb_ext.ExtImpMarsmedia
+	if err := json.Unmarshal(bidderExt.Bidder, &marsmediaExt); err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "ext.bidder.zone not provided",
+		}}
+	}
+
+	if marsmediaExt.ZoneID == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Zone is empty",
+		}}
+	}
+
+	validImpExists := false
+	for i := 0; i < len(request.Imp); i++ {
+		if request.Imp[i].Banner != nil {
+			bannerCopy := *requestIn.Imp[i].Banner
+			if bannerCopy.W == nil && bannerCopy.H == nil && len(bannerCopy.Format) > 0 {
+				firstFormat := bannerCopy.Format[0]
+				bannerCopy.W = &(firstFormat.W)
+				bannerCopy.H = &(firstFormat.H)
+				request.Imp[i].Banner = &bannerCopy
+				validImpExists = true
+			} else {
+				return nil, []error{&errortypes.BadInput{
+					Message: "No valid banner foramt in the bid request",
+				}}
+			}
+		} else if request.Imp[i].Video != nil {
+			validImpExists = true
+		}
+	}
+	if !validImpExists {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No valid impression in the bid request",
+		}}
+	}
+
+	request.AT = 1 //Defaulting to first price auction for all prebid requests
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Json not encoded",
```

Comment:
This error might not help much with debugging. Consider including the message from `err` in the string.